File: sandeep_reports/sandeep_reports/report/daily_business_summary/daily_business_summary.py
# Daily Business Summary — standard Script Report (auto-converted from safe_exec DB script)
import frappe


def execute(filters=None):
	filters = frappe._dict(filters or {})
	from_date = filters.get("from_date") or frappe.utils.nowdate()
	to_date   = filters.get("to_date")   or frappe.utils.nowdate()
	company   = filters.get("company")   or frappe.db.get_default("Company")

	columns = [
	    {"fieldname": "section",  "label": "Section",  "fieldtype": "Data",     "width": 450},
	    {"fieldname": "count",    "label": "Count",    "fieldtype": "Int",      "width": 80},
	    {"fieldname": "amount",   "label": "Amount",   "fieldtype": "Currency", "width": 200},
	    {"fieldname": "row_type", "label": "Type",     "fieldtype": "Data",     "hidden": 1},
	]

	params = {"fd": from_date, "td": to_date, "co": company}

	# ── Sales Invoices — net of credit notes ──────────────────────
	# is_return=1 means credit note (negative grand_total already stored negative)
	# SUM(grand_total) nets them off automatically
	sales = frappe.db.sql("""
	    SELECT
	        SUM(CASE WHEN is_return = 0 THEN 1 ELSE 0 END) AS cnt,
	        IFNULL(SUM(CASE WHEN is_return = 0 THEN grand_total ELSE 0 END), 0) AS total,
	        IFNULL(SUM(base_net_total), 0) AS net_total
	    FROM `tabSales Invoice`
	    WHERE docstatus = 1
	    AND posting_date BETWEEN %(fd)s AND %(td)s
	    AND company = %(co)s
	""", params, as_dict=True)[0]

	# ── Purchase Invoices — net of debit notes ────────────────────
	purchases = frappe.db.sql("""
	    SELECT
	        SUM(CASE WHEN is_return = 0 THEN 1 ELSE 0 END) AS cnt,
	        IFNULL(SUM(CASE WHEN is_return = 0 THEN grand_total ELSE 0 END), 0) AS total,
	        IFNULL(SUM(base_net_total), 0) AS net_total
	    FROM `tabPurchase Invoice`
	    WHERE docstatus = 1
	    AND posting_date BETWEEN %(fd)s AND %(td)s
	    AND company = %(co)s
	""", params, as_dict=True)[0]

	# ── Collections — Payment Entry ────────────────────────────────
	collections = frappe.db.sql("""
	    SELECT COUNT(*) AS cnt, IFNULL(SUM(paid_amount), 0) AS total
	    FROM `tabPayment Entry`
	    WHERE docstatus = 1
	    AND payment_type = 'Receive'
	    AND posting_date BETWEEN %(fd)s AND %(td)s
	    AND company = %(co)s
	""", params, as_dict=True)[0]

	# ── Payments — Payment Entry ───────────────────────────────────
	payments = frappe.db.sql("""
	    SELECT COUNT(*) AS cnt, IFNULL(SUM(paid_amount), 0) AS total
	    FROM `tabPayment Entry`
	    WHERE docstatus = 1
	    AND payment_type = 'Pay'
	    AND posting_date BETWEEN %(fd)s AND %(td)s
	    AND company = %(co)s
	""", params, as_dict=True)[0]

	# ── Collections — Journal Entry ────────────────────────────────
	je_recv = frappe.db.sql("""
	    SELECT COUNT(DISTINCT gle.voucher_no) AS cnt,
	           IFNULL(SUM(gle.debit), 0) AS total
	    FROM `tabGL Entry` gle
	    INNER JOIN `tabAccount` acc ON acc.name = gle.account
	    INNER JOIN `tabJournal Entry` je ON je.name = gle.voucher_no
	    WHERE gle.voucher_type = 'Journal Entry'
	    AND je.voucher_type IN ('Bank Entry', 'Cash Entry')
	    AND acc.account_type IN ('Bank', 'Cash')
	    AND gle.debit > 0
	    AND gle.is_cancelled = 0
	    AND gle.posting_date BETWEEN %(fd)s AND %(td)s
	    AND gle.company = %(co)s
	""", params, as_dict=True)[0]

	# ── Payments — Journal Entry ───────────────────────────────────
	je_pay = frappe.db.sql("""
	    SELECT COUNT(DISTINCT gle.voucher_no) AS cnt,
	           IFNULL(SUM(gle.credit), 0) AS total
	    FROM `tabGL Entry` gle
	    INNER JOIN `tabAccount` acc ON acc.name = gle.account
	    INNER JOIN `tabJournal Entry` je ON je.name = gle.voucher_no
	    WHERE gle.voucher_type = 'Journal Entry'
	    AND je.voucher_type IN ('Bank Entry', 'Cash Entry')
	    AND acc.account_type IN ('Bank', 'Cash')
	    AND gle.credit > 0
	    AND gle.is_cancelled = 0
	    AND gle.posting_date BETWEEN %(fd)s AND %(td)s
	    AND gle.company = %(co)s
	""", params, as_dict=True)[0]

	# ── Purchase Orders pending ────────────────────────────────────
	pending_po = frappe.db.sql("""
	    SELECT COUNT(*) AS cnt, IFNULL(SUM(grand_total), 0) AS total
	    FROM `tabPurchase Order`
	    WHERE docstatus = 1
	    AND status NOT IN ('Completed','Cancelled','Closed')
	    AND company = %(co)s
	""", {"co": company}, as_dict=True)[0]

	# ── Sales Orders pending ───────────────────────────────────────
	pending_so = frappe.db.sql("""
	    SELECT COUNT(*) AS cnt, IFNULL(SUM(grand_total), 0) AS total
	    FROM `tabSales Order`
	    WHERE docstatus = 1
	    AND status NOT IN ('Completed','Cancelled','Closed')
	    AND company = %(co)s
	""", {"co": company}, as_dict=True)[0]

	# ── Total Outstanding Receivable — Payment Ledger Entry ───────
	receivable = frappe.db.sql("""
	    SELECT IFNULL(SUM(amount), 0) AS total
	    FROM `tabPayment Ledger Entry`
	    WHERE party_type = 'Customer'
	    AND delinked = 0
	    AND posting_date <= %(td)s
	    AND company = %(co)s
	""", {"td": to_date, "co": company}, as_dict=True)[0]

	# ── Total Outstanding Payable — Payment Ledger Entry ─────────
	payable = frappe.db.sql("""
	    SELECT IFNULL(SUM(amount), 0) AS total
	    FROM `tabPayment Ledger Entry`
	    WHERE party_type = 'Supplier'
	    AND delinked = 0
	    AND posting_date <= %(td)s
	    AND company = %(co)s
	""", {"td": to_date, "co": company}, as_dict=True)[0]

	# ── Totals ─────────────────────────────────────────────────────
	# Use base_net_total for sales/purchase to match Sales Invoice Trends
	# base_net_total = taxable value (excl GST), same as what trends report shows
	s   = frappe.utils.flt(sales.get("net_total"))
	p   = frappe.utils.flt(purchases.get("net_total"))
	c   = frappe.utils.flt(collections.get("total"))
	py_ = frappe.utils.flt(payments.get("total"))
	jc  = frappe.utils.flt(je_recv.get("total"))
	jp  = frappe.utils.flt(je_pay.get("total"))
	net = (c + jc) - (py_ + jp)

	receivable_amt = frappe.utils.flt(receivable.get("total"))
	payable_amt    = abs(frappe.utils.flt(payable.get("total")))

	rows = [
	    # ── SALES ──────────────────────────────────────────────────
	    {"section": "─── SALES",            "count": None, "amount": None,           "row_type": "header"},
	    {"section": "Sales Invoices",        "count": frappe.utils.cint(sales.get("cnt")),      "amount": s,              "row_type": "sales"},
	    {"section": "Sales Orders Pending",  "count": frappe.utils.cint(pending_so.get("cnt")), "amount": frappe.utils.flt(pending_so.get("total")), "row_type": "info_so"},
	    {"section": "Total Receivable",      "count": None,                                      "amount": receivable_amt, "row_type": "total_recv"},

	    # ── PURCHASES ──────────────────────────────────────────────
	    {"section": "─── PURCHASES",          "count": None, "amount": None,         "row_type": "header"},
	    {"section": "Purchase Invoices",       "count": frappe.utils.cint(purchases.get("cnt")), "amount": p,            "row_type": "purchase"},
	    {"section": "Purchase Orders Pending", "count": frappe.utils.cint(pending_po.get("cnt")), "amount": frappe.utils.flt(pending_po.get("total")), "row_type": "info_po"},
	    {"section": "Total Payable",           "count": None,                                      "amount": payable_amt, "row_type": "total_pay"},

	    # ── CASH FLOW ───────────────────────────────────────────────
	    {"section": "─── CASH FLOW",         "count": None, "amount": None, "row_type": "header"},
	    {"section": "Collections (PE)",       "count": frappe.utils.cint(collections.get("cnt")), "amount": c,   "row_type": "collection"},
	    {"section": "Collections (JE)",       "count": frappe.utils.cint(je_recv.get("cnt")),      "amount": jc,  "row_type": "je_collection"},
	    {"section": "Payments (PE)",          "count": frappe.utils.cint(payments.get("cnt")),     "amount": py_, "row_type": "payment"},
	    {"section": "Payments (JE)",          "count": frappe.utils.cint(je_pay.get("cnt")),       "amount": jp,  "row_type": "je_payment"},
	    {"section": "Net Cash Flow",          "count": None,                                        "amount": net, "row_type": "net"},
	]

	report_summary = [
	    {"value": s,        "label": "Sales",       "datatype": "Currency", "currency": "INR", "indicator": "green"},
	    {"value": p,        "label": "Purchases",   "datatype": "Currency", "currency": "INR", "indicator": "red"},
	    {"value": c + jc,   "label": "Collections", "datatype": "Currency", "currency": "INR", "indicator": "green"},
	    {"value": py_ + jp, "label": "Payments",    "datatype": "Currency", "currency": "INR", "indicator": "red"},
	    {"value": net,      "label": "Net Cash",    "datatype": "Currency", "currency": "INR", "indicator": "blue"},
	]

	data = columns, rows, None, None, report_summary
	return data

File: sandeep_reports/sandeep_reports/report/daily_business_summary/daily_business_summary.js
frappe.query_reports["Daily Business Summary"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            reqd: 1,
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1,
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1,
        },
    ],

    onload: function (report) {
        if (document.getElementById("dbs-style")) return;
        var s = document.createElement("style");
        s.id = "dbs-style";
        s.textContent = `
            .dbs-wrap {
                display: grid;
                grid-template-columns: repeat(3, 1fr);
                gap: 14px;
                padding: 16px 0 4px;
            }
            @media (max-width: 900px) { .dbs-wrap { grid-template-columns: 1fr; } }

            .dbs-card { border: 1px solid var(--border-color); border-radius: 8px; overflow: hidden; background: var(--card-bg); }

            .dbs-head { padding: 10px 16px; font-size: 11px; font-weight: 700; letter-spacing: 1px; text-transform: uppercase; display: flex; align-items: center; gap: 6px; }
            .dbs-head.green { background: rgba(46,204,113,0.08); color: #27ae60; border-bottom: 2px solid #2ecc71; }
            .dbs-head.red   { background: rgba(231,76,60,0.08);  color: #c0392b; border-bottom: 2px solid #e74c3c; }
            .dbs-head.blue  { background: rgba(52,152,219,0.08); color: #2980b9; border-bottom: 2px solid #3498db; }

            .dbs-row { display: flex; justify-content: space-between; align-items: center; padding: 9px 16px; border-bottom: 1px solid var(--border-color); transition: background 0.15s; }
            .dbs-row.nav { cursor: pointer; }
            .dbs-row.nav:hover { background: var(--fg-hover-color, rgba(0,0,0,0.04)); }

            .dbs-sub-label { font-size: 10px; color: var(--text-light, #aaa); background: var(--control-bg); padding: 1px 5px; border-radius: 3px; margin-left: 4px; }

            .dbs-foot { display: flex; justify-content: space-between; align-items: center; padding: 9px 16px; background: var(--control-bg); border-top: 2px solid var(--border-color); transition: filter 0.15s; }
            .dbs-foot.nav { cursor: pointer; }
            .dbs-foot.nav:hover { filter: brightness(0.94); }

            .dbs-lbl { font-size: 12px; color: var(--text-muted); display: flex; align-items: center; gap: 5px; }
            .dbs-arr { font-size: 10px; color: #bbb; }
            .dbs-right { display: flex; align-items: baseline; gap: 7px; }
            .dbs-cnt { font-size: 10px; color: var(--text-muted); background: var(--control-bg); padding: 1px 6px; border-radius: 3px; font-weight: 600; }
            .dbs-val { font-size: 14px; font-weight: 700; font-family: monospace; }
            .dbs-val.green { color: #27ae60; }
            .dbs-val.red   { color: #c0392b; }
            .dbs-val.amber { color: #e67e22; }
            .dbs-val.blue  { color: #2980b9; }

            .dbs-foot-lbl { font-size: 11px; font-weight: 700; text-transform: uppercase; letter-spacing: 0.5px; color: var(--text-muted); display: flex; align-items: center; gap: 5px; }
            .dbs-foot-val { font-size: 15px; font-weight: 700; font-family: monospace; }
            .dbs-foot-val.green { color: #27ae60; }
            .dbs-foot-val.red   { color: #c0392b; }
            .dbs-foot-val.amber { color: #e67e22; }
            .dbs-foot-val.blue  { color: #2980b9; }

            .dbs-divider { height: 1px; background: var(--border-color); margin: 0 16px; opacity: 0.5; }
        `;
        document.head.appendChild(s);
    },

    after_datatable_render: function () {
        this.render_cards();
    },

    render_cards: function () {
        // Remove previous render
        var old = document.querySelector(".dbs-wrap");
        if (old) old.remove();

        var data = frappe.query_report.data || [];
        var get  = function (t) { return data.find(function (r) { return r.row_type === t; }) || {}; };

        var si  = get("sales");
        var so  = get("info_so");
        var rv  = get("total_recv");
        var pi  = get("purchase");
        var po  = get("info_po");
        var py_ = get("total_pay");
        var cl  = get("collection");
        var jc  = get("je_collection");
        var pm  = get("payment");
        var jp  = get("je_payment");
        var nt  = get("net");

        var company  = frappe.query_report.get_filter_value("company") || "";
        var fd       = frappe.query_report.get_filter_value("from_date") || "";
        var td       = frappe.query_report.get_filter_value("to_date") || "";
        var netColor = parseFloat(nt.amount || 0) >= 0 ? "blue" : "red";
        var dateRange = ["between", [fd, td]];

        function fmt(n) {
            if (n === null || n === undefined) return "—";
            n = parseFloat(n) || 0;
            var abs = Math.abs(n), sign = n < 0 ? "-" : "";
            if (abs >= 1e7) return sign + "₹" + (abs / 1e7).toFixed(2) + " Cr";
            if (abs >= 1e5) return sign + "₹" + (abs / 1e5).toFixed(2) + " L";
            if (abs >= 1e3) return sign + "₹" + (abs / 1e3).toFixed(1) + "K";
            return sign + "₹" + abs.toFixed(0);
        }

        var navHandlers = [];

        function row(label, sublabel, val, color, count, navFn) {
            var idx = navHandlers.length;
            navHandlers.push(navFn || null);
            var cls  = navFn ? " nav" : "";
            var arr  = navFn ? '<span class="dbs-arr">↗</span>' : '';
            var sub  = sublabel ? '<span class="dbs-sub-label">' + sublabel + '</span>' : '';
            var cnt  = count ? '<span class="dbs-cnt">' + count + '</span>' : '';
            return '<div class="dbs-row' + cls + '" data-dbs="' + idx + '">'
                + '<span class="dbs-lbl">' + label + sub + arr + '</span>'
                + '<div class="dbs-right">' + cnt
                + '<span class="dbs-val ' + color + '">' + fmt(val) + '</span>'
                + '</div></div>';
        }

        function foot(label, val, color, navFn) {
            var idx = navHandlers.length;
            navHandlers.push(navFn || null);
            var cls = navFn ? " nav" : "";
            var arr = navFn ? '<span class="dbs-arr">↗</span>' : '';
            return '<div class="dbs-foot' + cls + '" data-dbs="' + idx + '">'
                + '<span class="dbs-foot-lbl">' + label + arr + '</span>'
                + '<span class="dbs-foot-val ' + color + '">' + fmt(val) + '</span>'
                + '</div>';
        }

        // ── SALES ─────────────────────────────────────────────
        var salesHtml = '<div class="dbs-card">'
            + '<div class="dbs-head green">💹 Sales</div>'
            + row("Invoices", "", si.amount, "green", si.count, function () {
                frappe.route_options = { docstatus: 1, company: company, posting_date: dateRange };
                frappe.set_route("List", "Sales Invoice");
            })
            + row("Pending Orders", "", so.amount, "amber", so.count, function () {
                frappe.route_options = { docstatus: 1, company: company, status: "To Deliver and Bill" };
                frappe.set_route("List", "Sales Order");
            })
            + foot("Total Receivable", rv.amount, "amber", function () {
                frappe.route_options = { company: company };
                frappe.set_route("query-report", "Accounts Receivable-RBOL");
            })
            + '</div>';

        // ── PURCHASES ─────────────────────────────────────────
        var purchHtml = '<div class="dbs-card">'
            + '<div class="dbs-head red">🛒 Purchases</div>'
            + row("Invoices", "", pi.amount, "red", pi.count, function () {
                frappe.route_options = { docstatus: 1, company: company, posting_date: dateRange };
                frappe.set_route("List", "Purchase Invoice");
            })
            + row("Pending Orders", "", po.amount, "amber", po.count, function () {
                frappe.route_options = { docstatus: 1, company: company, status: "To Receive and Bill" };
                frappe.set_route("List", "Purchase Order");
            })
            + foot("Total Payable", py_.amount, "amber", function () {
                frappe.route_options = { company: company };
                frappe.set_route("query-report", "Accounts Payable-RBOL");
            })
            + '</div>';

        // ── CASH FLOW ─────────────────────────────────────────
        var cashHtml = '<div class="dbs-card">'
            + '<div class="dbs-head blue">💰 Cash Flow</div>'
            + row("Collections", "PE", cl.amount, "green", cl.count, function () {
                frappe.route_options = { docstatus: 1, company: company, payment_type: "Receive", posting_date: dateRange };
                frappe.set_route("List", "Payment Entry");
            })
            + row("Collections", "JE", jc.amount, "green", jc.count, function () {
                // No voucher_type filter — shows all Bank/Cash JEs for the date range
                frappe.route_options = { company: company, posting_date: dateRange };
                frappe.set_route("List", "Journal Entry");
            })
            + '<div class="dbs-divider"></div>'
            + row("Payments Out", "PE", pm.amount, "red", pm.count, function () {
                frappe.route_options = { docstatus: 1, company: company, payment_type: "Pay", posting_date: dateRange };
                frappe.set_route("List", "Payment Entry");
            })
            + row("Payments Out", "JE", jp.amount, "red", jp.count, function () {
                // No voucher_type filter — shows all Bank/Cash JEs for the date range
                frappe.route_options = { company: company, posting_date: dateRange };
                frappe.set_route("List", "Journal Entry");
            })
            + foot("Net Cash Flow", nt.amount, netColor, null)
            + '</div>';

        // Inject before datatable
        var dt = document.querySelector(".datatable");
        if (dt) {
            dt.insertAdjacentHTML("beforebegin",
                '<div class="dbs-wrap">' + salesHtml + purchHtml + cashHtml + '</div>');
        }

        // Attach click handlers
        document.querySelectorAll("[data-dbs]").forEach(function (el) {
            var fn = navHandlers[parseInt(el.getAttribute("data-dbs"), 10)];
            if (fn) el.addEventListener("click", fn);
        });

        // ── Hide the raw table completely ──────────────────────
        var hide = [
            document.querySelector(".datatable"),
            document.querySelector(".dt-wrapper"),
            document.querySelector(".report-wrapper .datatable"),
        ];
        hide.forEach(function (el) {
            if (el && el.parentElement) el.parentElement.style.display = "none";
        });
    },
};